File: rotate-array/rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        k = k % nums.length;
        if(k > 0) {
            reverse(nums,0,nums.length-1-k);
            reverse(nums,nums.length-k, nums.length-1);
            reverse(nums,0,nums.length-1);
        }
    }

    public void reverse(int[] nums, int start, int end) {
        while(start < end) {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
}

File: 189-rotate-array/rotate-array.py
class Solution:
    def reverse(self, nums, start, end):
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
            start += 1
            end -= 1

    def rotate(self, nums: List[int], k: int) -> None:
        k = k % len(nums)
        if k:
            self.reverse(nums, 0, len(nums)-1)
            self.reverse(nums, 0, k-1)
            self.reverse(nums, k, len(nums)-1)